File: 1334_Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        dist = [[float('inf')] * n for _ in range(n)]
        res = smallest = n
        for e1, e2, w in edges:
            dist[e1][e2] = dist[e2][e1] = w
        for i in range(n):
            dist[i][i] = 0
        for k in range(n):
            for i in range(n):
                for j in range(i):
                    dist[i][j] = dist[j][i] = min(dist[i][j], dist[i][k] + dist[k][j])
        for i in range(n):
            count = 0
            for j in range(n):
                if dist[i][j] <= distanceThreshold:
                    count += 1
            if 0 < count <= smallest:
                res, smallest = i, count
        return res

File: 1334_Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance.cpp
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        // Floyd-Warshall
        vector<vector<int>> dist (n, vector<int> (n, INT_MAX));
        for (vector<int> edge : edges) {
            dist[edge[0]][edge[1]] = edge[2];
            dist[edge[1]][edge[0]] = edge[2];
        }
        for (int k = 0; k < n; ++k) {
            dist[k][k] = 0;
            for (int i = 0; i < n; ++i) {
                for (int j = 0; j < n; ++j) {
                    if (dist[i][k] == INT_MAX || dist[j][k] == INT_MAX)
                        continue;
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }

        int minimal_adjs = n, ans = -1;
        for (int i = 0; i < n; ++i) {
            int adjs = 0;
            for (int j = 0; j < n; ++j) {
                if (dist[i][j] <= distanceThreshold) adjs++;
            }
            if (adjs <= minimal_adjs) {
                minimal_adjs = adjs;
                ans = max(ans, i);
            }
        }

        return ans;
    }
};
